x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResolveFieldsAction.java

```diff
@@ -6,59 +6,354 @@
+                    listener.map(caps -> new EsqlResolveFieldsResponse(caps, minTransportVersion.get()))
+                );
+            }
+        })) {
+            // local cluster
+            final RequestDispatcher requestDispatcher = new RequestDispatcher(
+                clusterService,
+                transportService,
+                projectResolver,
+                indicesService.getCoordinatorRewriteContextProvider(() -> nowInMillis),
+                task,
+                request,
+                localIndices,
+                nowInMillis,
+                concreteIndices,
+                singleThreadedExecutor,
+                handleIndexResponse,
+                handleIndexFailure,
+                refs.acquire()::close
+            );
+            requestDispatcher.execute();
+
+            // this is the cross cluster part of this API - we force the other cluster to not merge the results but instead
+            // send us back all individual index results.
+            for (Map.Entry<String, OriginalIndices> remoteIndices : remoteClusterIndices.entrySet()) {
+                String clusterAlias = remoteIndices.getKey();
+                OriginalIndices originalIndices = remoteIndices.getValue();
+                FieldCapabilitiesRequest remoteRequest = TransportFieldCapabilitiesAction.prepareRemoteRequest(
+                    clusterAlias,
+                    request,
+                    originalIndices,
+                    nowInMillis
+                );
+                ActionListener<EsqlResolveFieldsResponse> remoteListener = ActionListener.wrap(response -> {
+                    for (FieldCapabilitiesIndexResponse resp : response.caps().getIndexResponses()) {
+                        String indexName = RemoteClusterAware.buildRemoteIndexName(clusterAlias, resp.getIndexName());
+                        handleIndexResponse.accept(
+                            new FieldCapabilitiesIndexResponse(
+                                indexName,
+                                resp.getIndexMappingHash(),
+                                resp.get(),
+                                resp.canMatch(),
+                                resp.getIndexMode()
+                            )
+                        );
+                    }
+                    for (FieldCapabilitiesFailure failure : response.caps().getFailures()) {
+                        Exception ex = failure.getException();
+                        for (String index : failure.getIndices()) {
+                            handleIndexFailure.accept(RemoteClusterAware.buildRemoteIndexName(clusterAlias, index), ex);
+                        }
+                    }
+                    if (response.minTransportVersion() == null) {
+                        minTransportVersion.set(null);
+                    } else {
+                        minTransportVersion.accumulateAndGet(response.minTransportVersion(), TransportVersion::min);
```

Comment:
We may hit NullPointerException if we receive a response from an old cluster then another response from a new cluster. Also, can we handle both branches within a single accumulateAndGet call to avoid data races when processing two responses concurrently?